### ISSUE TITLE
Implement string interpolation via manipulating StringBuilder directly instead of emitting String.Format()

### ISSUE THREAD

**Issue by HellBrick** [dotnet/roslyn#194]
#### Current situation
At the moment string interpolation is usually compiled into a `String.Format()` call like this:

```C#
string s = $"[{TimeStamp}] {Level} - {Logger} - {Message}";
```
->
```C#
string s = string.Format("[{0}] {1} - {2} - {3}", new object[] {
            this.TimeStamp,
            this.Level,
            this.Logger,
            this.Message
        });
```

This leads to allocating an array of arguments (if there are more than 3 of them), boxing all primitive types and parsing the format string, which all seems like a waste of time, because the compiler already has all the information to emit the efficient code to build the string.

#### Proposition

The idea is to emit the code that manipulates `StringBuilder` directly instead of `String.Format()` call. For the example that I've mentioned it would look something like this:

```C#
// The same capacity estimation that String.Format() does can be used here,
// but with the value calculated during the compilation
StringBuilder _builder_s = StringBuilderCache.Acquire( 21 + 4 * 8 ); 
_builder_s.Append( "[" );
_builder_s.Append( this.TimeStamp );
_builder_s.Append( "] " );
_builder_s.Append( this.Level );
_builder_s.Append( " - " );
_builder_s.Append( this.Logger );
_builder_s.Append( " - " );
_builder_s.Append( this.Message );
string s = StringBuilderCache.GetStringAndRelease( _builder_s );
```

This seems like a fairly straightforward optimization to implement. The only problem that I see at the first glance is that `StringBuilderCache` is internal; however, there's probably a way to work around that for the code generated by compiler that I'm simply not aware of.

**Comment by RichiCoder1** [comment]
How does this deal with cases where there's actually formatting occurring? Would it just fall back to string.Format (and IFormattable) where necessary?

**Comment by HellBrick** [comment]
It would be similar to the way `StringBuilder.AppendFormat()` works: if the type of the argument implements `IFormattable`, call `ToString( format )`, otherwise call `ToString()`:

```C#
string str = $"{DateTime.UtcNow :o}{Task.FromResult( 42 ) :g2}";
```
->
```C#
StringBuilder builder = StringBuilderCache.Acquire( /* capacity */ );
// Format string specified + IFormattable => use the format string
builder.Append( DateTime.UtcNow.ToString( "o" ) );
// Format specified, but no IFormattable => ignore the format string
builder.Append( Task.FromResult( 42 ).ToString() );
string str = StringBuilderCache.GetStringAndRelease( builder );
```

**Comment by alrz** [comment]
For something like `$"{b:x2}"` this can be just translate to `b.ToString("x2")`.

**Comment by rhencke** [comment]
What about using String.Concat instead?  It already handles allocation efficiently by summing the length of all strings passed to it before allocating the array: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/String.cs#L3088

**Comment by alrz** [comment]
@rhencke still needs to allocate an array.

**Comment by HaloFour** [comment]
Remember that interpolation allows specifying the alignment.  In those cases calling `IFormattable.ToString` or `ICustomFormatter.ToString` wouldn't be sufficient.  Could always fallback to `StringBuilder.AppendFormat` for that case since there is an overload that accepts a single arg and wouldn't require an array allocation.

```cs
var name = "Bob";
var s = $"Hello {name,-10}";

StringBuilder builder = StringBuilderCache.Acquire( /* capacity */ );
builder.Append("Hello ");
builder.AppendFormat("{0,-10}", name);
string str = StringBuilderCache.GetStringAndRelease(builder);
```

**Comment by rhencke** [comment]
@alrz Ahh.  I see how it's working now.  Thank you.

**Comment by bbarry** [comment]
This isn't as straightforward as it appears. You have to call the `.AppendFormat()` method regardless of whether alignment or a format string specifier is present. I think it might be safe to skip for `string` arguments, but for everything else you must at least attempt to cast to `IFormattable` and potentially deal with alignments.

The example:

    string s = $"[{TimeStamp}] {Level} - {Logger} - {Message}";

needs to be:

    string s = StringBuilderCache.GetStringAndRelease(
            StringBuilderCache.Acquire( /* capacity */ )
                .Append("[")
                .AppendFormat("{0}", TimeStamp)
                .Append("] ")
                .AppendFormat("{0}", Level)
                .Append(" - ")
                .AppendFormat("{0}", Logger)
                .Append(" - ")
                .AppendFormat("{0}", Message)
        );

I think you can safely skip `.AppendFormat()` for `string` types that have no alignment attached as they are sealed and do not implement `IFormattable` (and it would be a breaking change to do so). However, most other built in types do. It isn't safe to just call the public overload on most primitives because there is no specification that requires `IFormattable` to be public, so instead you must safe cast to `IFormattable`, and once you have an `IFormattable` you might as well handle the case of having a format specifier: `(arg as IFormattable)?.ToString(fmt, null) ?? arg?.ToString() ?? ""`. All to handle the base case of not needing to call `.AppendFormat()` when you have no format string specifier (you get handling the specifier almost for free).

So you could maybe be:

    DateTime _0 = TimeStamp;
    LogLevel _1 = Level; //assuming this is a class in a library somewhere
    ILogger _2 = Logger;
    string _3 = Messsage;
    string s = StringBuilderCache.GetStringAndRelease(
            StringBuilderCache.Acquire( /* capacity */ )
                .Append("[")
                .Append((_0 as IFormattable)?.ToString(null, null))
                .Append("] ")
                .Append((_1 as IFormattable)?.ToString(null, null) ?? _1?.ToString())
                .Append(" - ")
                .Append((_2 as IFormattable)?.ToString(null, null) ?? _2?.ToString())
                .Append(" - ")
                .Append(_3)
        );


Relying on that code fragment never having to change for any possible future revision of .NET is probably not a good idea either. So, back to `.AppendFormat()` for all holes in the interpolated strings... 

Doing some testing on potential changes that could be done ([test code](https://gist.github.com/bbarry/4b25a2ae1ab87a425825)), using `StringBuilder` directly over `string.Format` provides very negligible performance improvements (there are a few cases where not allocating the array for the call `string.Format(string, params object[])` call results in a faster computation, but this has diminishing returns as the array length increases, or 0 difference anyway if the cache misses). 

**TLDR**: I think there is very little point in making this change.

The enhancement here has some pros:

* avoids an array allocation 
* *might* be able to avoid calling `.AppendFormat()` on certain well known built in types (namely `string`) under specific circumstances (no alignments)

But it also has some cons:

* significant code size increases for ambiguous performance improvements
* still must call `.AppendFormat()` most of the time (which must still process over a format argument)
* still boxes arguments
* some complexity needed to access `StringBuilderCache` (BCL change?)

:-1:

If there was some BCL change that created some helper methods for string interpolation, specifically a method or methods that interpolated a string that was a single hole: `$"{...}"` and avoided boxing on that method, I think it would be worth using in either a `string.Concat` or `StringBuilder` implementation. Alternatively an API addition to `StringBuilder` to add an `.Append<T>(T item, string format, int align)` overload, might be worth utilizing.
